provisioning/transport/http/src/http.ts

```diff
@@ -0,0 +1,251 @@
+  constructor(idScope: string, registrationId: string, httpBase?: Base) {
+    super();
+    this._idScope = idScope;
+    this._registrationId = registrationId;
+
+    this._fsm = new machina.Fsm({
+      namespace: 'http-provisioning',
+      initialState: 'idle',
+      states: {
+        idle: {
+          _onEnter: (callback, err, response) => {
+            this._pollingTimer = null;
+            this._restApiClient = null;
+            this._body = null;
+            this._forceRegistration = false;
+            this._operationId = '';
+            this._registrationCallback = null;
+            if (callback) {
+              callback(err, response);
+            }
+          },
+          register: (authorization, forceRegistration, body, callback) => {
+            this._registrationCallback = callback;
+            this._forceRegistration = forceRegistration;
+            this._body = body;
+            if (authorization.hasOwnProperty('cert')) {
+              this._restApiClient = new RestApiClient({ 'host' : this._provisioningServer , 'x509' : authorization}, 'azure-node-provisioning', httpBase);
+            } else {
+              this._restApiClient = new RestApiClient({ 'host' : this._provisioningServer , 'sharedAccessSignature' : authorization}, 'azure-node-provisioning', httpBase);
+            }
+            this._fsm.transition('registering', callback);
+          },
+          disconnect : (callback) => {
+            // nothing to do.
+            callback();
+          }
+        },
+        registering: {
+          _onEnter: (callback) => {
+            this._sendRegistrationRequest((err, responseBody) => {
+              if (this._registrationInProgress()) { // make sure we weren't cancelled before doing something with the response
+                if (err) {
+                    this._fsm.transition('idle', callback, err);
+                } else if (responseBody.status.toLowerCase() === 'assigning') {
+                  /* Codes_SRS_NODE_PROVISIONING_HTTP_18_016: [ If the registration response has a success code with a 'status' of 'Assigning', `register` shall fire an `operationStatus` event with the deserialized body ] */
+                  this.emit('operationStatus', responseBody);
+                  /* Codes_SRS_NODE_PROVISIONING_HTTP_18_017: [ If the registration response has a success code with a 'status' of 'Assigning', `register` shall start polling for operation updates ] */
+                  this._operationId = responseBody.operationId;
+                  this._fsm.transition('waiting_to_poll', callback);
+                } else if (responseBody.status.toLowerCase() === 'assigned') {
+                  /* Codes_SRS_NODE_PROVISIONING_HTTP_18_015: [ If the registration response has a success code with a 'status' of 'Assigned', `register` call the `callback` with `err` == `null` and result `containing` the deserialized body ] */
+                  this._fsm.transition('idle', callback, null, responseBody);
```

Comment:
>idle [](start = 40, length = 4)

so there is no "registered" state? we don't have a use for it? registration operations are idempotent right?

Reply:
no registered state.  need to double check.  requests are not idempotent when forceRegistration is used.

---
In reply to: [142298471](https://github.com/Azure/azure-iot-sdk-node/pull/99#discussion_r142298471) [](ancestors = 142298471)